R/site.R

```diff
@@ -0,0 +1,191 @@
+                        ...) {
+
+  # normalize input (capture original input first)
+  original_input <- input
+  input <- input_as_dir(input)
+
+  # find the site generator
+  generator <- site_generator(input, output_format, encoding)
+  if (is.null(generator))
+    stop("No index.Rmd with site entry found.")
+
+  # execute it
+  generator$render(output_format = output_format,
+                   quiet = quiet,
+                   encoding = encoding)
+
+  # print the name of the index file (for RStudio Preview)
+  if (!quiet) {
+    # if the input was a filename use that as a base (otherwise
+    # use index.Rmd)
+    if (!dir_exists(original_input))
+      output <- file_with_ext(basename(original_input), "html")
+    else
+      output <- "index.html"
+    output <- file.path(input, generator$output_dir, output)
+    message("\nOutput created: ", relative_to(getwd(), output))
+  }
+}
+
+#' @rdname render_site
+#' @export
+site_generator <- function(input = ".",
+                           output_format = NULL,
+                           encoding = getOption("encoding")) {
+
+  # normalize input
+  input <- input_as_dir(input)
+
+  # if we have an index.Rmd then check it's yaml for "site:"
+  index <- file.path(input, "index.Rmd")
+  if (file.exists(index)) {
+
+    # read index.Rmd and extract the front matter
+    index_lines <- read_lines_utf8(index, encoding)
+    front_matter <- parse_yaml_front_matter(index_lines)
+    if (!is.null(front_matter) && !is.null(front_matter$site)) {
+
+      # discover the site generator function (if "site" is a simple true
+      # value then we use the default output format's name + _site). If
+      # that function doesn't exist then just use rmarkdown::default_site,
+      # which will just render all of the Rmd files in the directory
+      if (isTRUE(front_matter$site)) {
+        if (is.null(output_format))
+          output_format <- default_output_format(index, encoding)$name
+        site_function_name = paste0(output_format, "_site")
+        create_site_generator <- tryCatch(
+                                   eval(parse(text = site_function_name)),
+                                   error = function(x) NULL)
+        if (is.null(create_site_generator))
+          create_site_generator <- rmarkdown::default_site;
+      } else {
+        create_site_generator <- eval(parse(text = front_matter$site))
+      }
+
+      # execute the create_site_generator function
+      create_site_generator(input)
+
+    } else {
+      NULL  # no site: front matter
+    }
+  } else {
+    NULL # no index.Rmd
+  }
+}
+
+#' @rdname render_site
+#' @export
+default_site <- function(input, ...) {
+
+  # define render function (use ... to gracefully handle future args)
+  render <- function(output_format = NULL,
+                     quiet = FALSE,
+                     encoding = getOption("encoding"),
+                     ...) {
+    files <- list.files(input, pattern = glob2rx("*.Rmd"))
```

Comment:
How does this approach interact with the various ways folks may include Rmd inside other Rmd? I don't know if this is the perfect example or all that common, but in just a couple moments of poking around, I found:
http://stackoverflow.com/a/25979000

I also wonder if `default_site` should worry about directory traversal rather than just looking at the current-directory. Although perhaps this is best left to an alternate site generator.